### PR TITLE
ProtoData-based validation

### PR DESCRIPTION
**Pull request by dmdashenkov**
Introducing the skeleton for validation code generation on ProtoData.

In this PR we introduce a view for language-agnostic validation rules and a Java renderer that generated validation code of arbitrary complexity.

There is also a policy that generates required field validation rules based on the `(required)` option from `spine/options.proto`.

This is a trial run for ProtoData. In future, these components should be migrated to a permanent location in a separate repo under `SpineEventEngine` org.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
There is a known [issue](https://github.com/SpineEventEngine/ProtoData/issues/13) with the current implementation.

Resolving the issue will probably require us to fork Google Truth and offer them a small change. Alternatively, it may require a small conceptual change in Java validation.

To decide which path to take, I suggest we convene and discuss.

**Comment by dmdashenkov** [comment]
@armiol, @alexander-yevsyukov, PTAL.

File: protodata/src/main/java/io/spine/kanban/codegen/SpineOptionsProvider.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+package io.spine.kanban.codegen;
+
+import com.google.protobuf.ExtensionRegistry;
+import io.spine.option.OptionsProto;
+import io.spine.protodata.option.OptionsProvider;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * An {@link OptionsProvider} which registers Spine options from {@code options.proto}.
```
Comment (armiol) [inline protodata/src/main/java/io/spine/kanban/codegen/SpineOptionsProvider.java:35]: I assume, "... in ProtoData's <something>."?

What I mean is that one may wonder why register? These options seem to be registered already, as the Protobuf compiler is able to find them perfectly...

File: protodata/src/main/java/io/spine/kanban/codegen/ValidateInsertionPointPrinter.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+
+/**
+ * An {@link InsertionPointPrinter} which adds the {@link Validate} point to all the message types
+ * which have an associated {@link MessageValidation} view.
+ */
+@SuppressWarnings("unused") // Accessed via reflection by ProtoData.
+public final class ValidateInsertionPointPrinter extends InsertionPointPrinter {
```
Comment (armiol) [inline protodata/src/main/java/io/spine/kanban/codegen/ValidateInsertionPointPrinter.java:44]: I would call it `PrintValidationInsertionPoints`.

And as for its method `getSupportedInsertionPoints`, I would rename it to `supportedPoints()`.

File: protodata/src/main/java/io/spine/kanban/codegen/ValidationPlugin.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * A ProtoData plugin which attaches validation-related policies and views.
+ */
+@SuppressWarnings("unused") // Loaded by ProtoData via reflection.
+public class ValidationPlugin implements Plugin {
+
+    @NotNull
+    @Override
+    public Set<Policy<?>> getPolicies() {
```
Comment (armiol) [inline protodata/src/main/java/io/spine/kanban/codegen/ValidationPlugin.java:45]: Probably, we'll need to discuss vocally why our getters start with `get`. I suspect there is a reason...

File: protodata/src/main/java/io/spine/kanban/codegen/package-info.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * Contains the ProtoData validation plugin which generates Java code.
+ */
+
```
Comment (armiol) [inline protodata/src/main/java/io/spine/kanban/codegen/package-info.java:30]: Please kill this line, as we are documenting the `package`.

File: protodata/src/main/kotlin/io/spine/kanban/codegen/ErrorMessage.kt [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        if (field != null) {
+            violationBuilder = violationBuilder.chainSet("field_path", pathFrom(field))
+        }
+        if (fieldValue != null) {
+            violationBuilder = violationBuilder.chainSet("field_value", fieldValue.packToAny())
+        }
+        return violationBuilder.chainBuild()
+    }
+
+    private fun pathFrom(field: Field): Expression {
```
Comment (armiol) [inline protodata/src/main/kotlin/io/spine/kanban/codegen/ErrorMessage.kt:137]: Isn't it `pathOf()`? I mean, a path from <some field> should probably a path to somewhere. Taken by someone who wants to get from this field to some particular place.

File: protodata/src/main/kotlin/io/spine/kanban/codegen/Expression.kt [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+    override fun toString(): String {
+        return "FieldAccess[$message#${name.value}]"
+    }
+}
+
+/**
+ * An expression of a Java method call.
+ *
+ * Might be a static or an instance method. In the case of the former, the receiver expression
```
Comment (armiol) [inline protodata/src/main/kotlin/io/spine/kanban/codegen/Expression.kt:266]: I think it is "May be...".

File: protodata/src/main/kotlin/io/spine/kanban/codegen/JavaCodeGenerator.kt [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    fun condition(): Expression
+
+    /**
+     * Forms an error message for the found violation.
+     */
+    fun error(): ErrorMessage
+}
+
+/**
+ * A generator for code which checks a simple one field rule.
```
Comment (armiol) [inline protodata/src/main/kotlin/io/spine/kanban/codegen/JavaCodeGenerator.kt:102]: "one-field"?

File: protodata/src/main/proto/spine/validation/rule.proto [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+//
+message Rule {
+
+    // The field which is being compared.
+    protodata.Field field = 1;
+
+    oneof operation {
+
+        // A simple comparison sign.
+        Sign sign = 2;
```
Comment (armiol) [inline protodata/src/main/proto/spine/validation/rule.proto:30]: To me, this is an `ComparisonOperator`. `Sign` is too wide to me, as "-7" is also a sign and a number, while it's clear it's far from comparison.

File: protodata/src/main/proto/spine/validation/rule.proto [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    LESS_THAN = 3;
+
+    GREATER_THAN = 4;
+
+    LESS_OR_EQUAL = 5;
+
+    GREATER_OR_EQUAL = 6;
+
+    DISTINCT = 7;
```
Comment (armiol) [inline protodata/src/main/proto/spine/validation/rule.proto:67]: Following up my previous comment, this is clearly not a sign.

File: protodata/src/main/proto/spine/validation/rule.proto [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// A non-predefined comparison option.
```
Comment (armiol) [inline protodata/src/main/proto/spine/validation/rule.proto:70]: I would speak of it as of `CustomOperator`. And also, I would avoid saying "option" here, in the domain so dangerously close to Protobuf.

File: protodata/src/main/proto/spine/validation/rule.proto [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+message UnknownOperation {
+
+    // Describes how the renderer should compare the values.
+    string description = 1;
+}
+
+message RuleOrComposite {
```
Comment (armiol) [inline protodata/src/main/proto/spine/validation/rule.proto:77]: I would name this `Rule`. Then its children would be `PlainRule` and `CompositeRule`.

Comment (armiol) [inline protodata/src/main/proto/spine/validation/rule.proto:77]: I would name this `Rule`. Then its children would be `PlainRule` and `CompositeRule`.

File: protodata/src/main/proto/spine/validation/rule.proto [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    //
+    // May include tokens `{left}`, `{right}`, and `{operation}`, which will be replaced with at
+    // runtime with the error messages of the `left` and `right` rules and the words "and" or "or"
+    // or "xor".
+    //
+    string error_message = 4;
+}
+
+// A boolean operator.
+enum BinaryOperation {
```
Comment (armiol) [inline protodata/src/main/proto/spine/validation/rule.proto:113]: Similarly to the `Sign`-related comments, this one seems `LogicalOperator`.

**Review by armiol — CHANGES_REQUESTED** [review]
@dmdashenkov I really like what I see in the "client code"-part of this PR! However, in the "library" part of this changeset, there are some things I would improve. Please see my comments.

File: protodata/src/main/kotlin/io/spine/kanban/codegen/Expression.kt (hunk 6 above)
Comment (dmdashenkov) [reply protodata/src/main/kotlin/io/spine/kanban/codegen/Expression.kt]: Changed to "Can be...". The sentence describes the physical ability, not a permission.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

**Comment by dmdashenkov** [comment]
The build is red right now as this iteration relies on [a version of ProtoData](https://github.com/SpineEventEngine/ProtoData/pull/14) which is not yet published.

File: protodata/src/main/proto/spine/validation/rule.proto [hunk 13]
```diff
@@ -19,18 +19,18 @@ import "google/protobuf/any.proto";
...
 
     oneof operation {
 
         // A simple comparison sign.
-        Sign sign = 2;
+        ComparisonOperator sign = 2;
 
         // A custom comparison method, which cannot be covered by a combination of `Sign`s.
-        UnknownOperation unknown_operation = 3;
+        CustomOperator unknown_operation = 3;
```
Comment (armiol) [inline protodata/src/main/proto/spine/validation/rule.proto:33]: I believe, with the `UnknownOperation` renamed to `CustomOperator` this field is now `custom_operator`?

File: protodata/src/main/proto/spine/validation/rule.proto [hunk 14]
```diff
@@ -92,13 +92,13 @@ message RuleOrComposite {
 message CompositeRule {
 
     // First rule.
-    RuleOrComposite left = 1;
+    Rule left = 1;
 
     // The operation which joins two rules.
-    BinaryOperation operation = 2;
+    LogicalOperator operation = 2;
```
Comment (armiol) [inline protodata/src/main/proto/spine/validation/rule.proto:98]: I would rename this field to `operator` and adjust its description.

**Review by armiol — APPROVED** [review]
@dmdashenkov LGTM with two comments.

**Pull request merged**